File: javascript/p0239.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var maxSlidingWindow = function (nums, k) {
  const ans = []; // Stores the max of each window
  const maxToMin = []; // Deque to keep track of max elements

  let left = 0,
    right = 0; // Window pointers

  // Process the first window
  while (right < k) {
    while (maxToMin.length && maxToMin[maxToMin.length - 1] < nums[right]) {
      maxToMin.pop(); // Maintain decreasing order in deque
    }
    maxToMin.push(nums[right]);
    right++;
  }

  ans.push(maxToMin[0]); // First window max

  // Process the rest of the windows
  while (right < nums.length) {
    if (maxToMin[0] === nums[left]) {
      maxToMin.shift(); // Remove element leaving the window
    }
    while (maxToMin.length && maxToMin[maxToMin.length - 1] < nums[right]) {
      maxToMin.pop();
    }
    maxToMin.push(nums[right]);
    ans.push(maxToMin[0]); // Current window max
    left++;
    right++;
  }

  return ans; // Return all window maximums
};

// Exemple usage
console.log(maxSlidingWindow([1, 3, -1, -3, 5, 3, 6, 7], 3));

File: python/p0239.py
from collections import deque  # Efficient deque for popping from both ends
from typing import List  # For type hinting

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        ans = []  # Stores the max of each window
        max_to_min = deque()  # Deque to keep track of max elements

        left = right = 0  # Window pointers

        # Process the first window
        while right < k:
            while max_to_min and max_to_min[-1] < nums[right]:
                max_to_min.pop()  # Maintain decreasing order in deque
            max_to_min.append(nums[right])
            right += 1

        ans.append(max_to_min[0])  # First window max

        # Process the rest of the windows
        while right < len(nums):
            if max_to_min[0] == nums[left]:
                max_to_min.popleft()  # Remove element leaving the window
            while max_to_min and max_to_min[-1] < nums[right]:
                max_to_min.pop()
            max_to_min.append(nums[right])
            ans.append(max_to_min[0])  # Current window max
            left += 1
            right += 1

        return ans  # Return all window maximums

# Example usage
solution = Solution()
print(solution.maxSlidingWindow([1,3,-1,-3,5,3,6,7], 3))